db/src/db.rs

```diff
@@ -416,45 +437,438 @@ impl DB {
+    pub fn resolve_avs<'a>(&self, conn: &rusqlite::Connection, avs: &'a [AVPair]) -> Result<AVMap<'a>> {
+        // Start search_id's at some identifiable number.
+        let initial_search_id = 2000;
+        let values_per_statement = 4;
+
+        // We map [a v] -> numeric search_id -> e, and then we use the search_id lookups to finally
+        // produce the map [a v] -> e.
+        //
+        // TODO: `collect` into a HashSet so that any (a, v) is resolved at most once.
+        let chunks: itertools::IntoChunks<_> = avs.into_iter().enumerate().chunks(::SQLITE_MAX_VARIABLE_NUMBER / 4);
+
+        // We'd like to `flat_map` here, but it's not obvious how to `flat_map` across `Result`.
+        // Alternatively, this is a `fold`, and it might be wise to express it as such.
+        let results: Result<Vec<Vec<_>>> = chunks.into_iter().map(|chunk| -> Result<Vec<_>> {
+            let mut count = 0;
+
+            // We must keep these computed values somewhere to reference them later, so we can't
+            // combine this `map` and the subsequent `flat_map`.
+            let block: Vec<(i64, i64, ToSqlOutput<'a>, i32)> = chunk.map(|(index, &(a, ref v))| {
+                count += 1;
+                let search_id: i64 = initial_search_id + index as i64;
+                let (value, value_type_tag) = v.to_sql_value_pair();
+                (search_id, a, value, value_type_tag)
+            }).collect();
+
+            // `params` reference computed values in `block`.
+            let params: Vec<&ToSql> = block.iter().flat_map(|&(ref searchid, ref a, ref value, ref value_type_tag)| {
+                // Avoid inner heap allocation.
+                once(searchid as &ToSql)
+                    .chain(once(a as &ToSql)
+                           .chain(once(value as &ToSql)
+                                  .chain(once(value_type_tag as &ToSql))))
+            }).collect();
+
+            // TODO: cache these statements for selected values of `count`.
+            let values: String = repeat_values(values_per_statement, count);
+            let s: String = format!("WITH t(search_id, a, v, value_type_tag) AS (VALUES {}) SELECT t.search_id, d.e \
+                                     FROM t, all_datoms AS d \
+                                     WHERE d.index_avet IS NOT 0 AND d.a = t.a AND d.value_type_tag = t.value_type_tag AND d.v = t.v", values);
+            let mut stmt: rusqlite::Statement = conn.prepare(s.as_str())?;
+
+            let m: Result<Vec<(i64, Entid)>> = stmt.query_and_then(&params, |row| -> Result<(i64, Entid)> {
+                Ok((row.get_checked(0)?, row.get_checked(1)?))
+            })?.collect();
+            m
+        }).collect::<Result<Vec<Vec<(i64, Entid)>>>>();
+
+        // Flatten.
+        let results: Vec<(i64, Entid)> = results?.as_slice().concat();
+
+        // Create map [a v] -> e.
+        let m: HashMap<&'a AVPair, Entid> = results.into_iter().map(|(search_id, entid)| {
```

Comment:
Yeah, there's a bunch of ways to phrase this.  I want it this way to demonstrate the chunking technique with higher order functions (partly for myself!).  I'm going to keep it as it is for now, but I've filed https://github.com/mozilla/mentat/issues/262 to do better eventually.